### PR TITLE
Make archive session extraction interruptible

### PR DESCRIPTION
**Pull request by dominiklohmann**

### CODE REVIEW

File: libvast/src/system/archive.cpp [hunk 1]
```diff
@@ -77,94 +77,124 @@ archive(archive_type::stateful_pointer<archive_state> self, path dir,
   });
   self->set_down_handler([=](const down_msg& msg) {
     VAST_DEBUG(self, "received DOWN from", msg.source);
-    self->state.active_exporters.erase(msg.source);
+    auto& st = self->state;
+    st.active_exporters.erase(msg.source);
+    while (st.unhandled_ids.front().second == msg.source)
```
Comment (mavam) [inline libvast/src/system/archive.cpp:82]: When using `.front()`, please add an assertion that the container is not empty.

File: libvast/src/system/archive.cpp [hunk 2]
```diff
@@ -77,94 +77,124 @@ archive(archive_type::stateful_pointer<archive_state> self, path dir,
...
-                == 0) {
-              VAST_DEBUG(self, "dismisses query for inactive sender");
-              return make_error(ec::no_error);
+  return {
+    [=](const ids& xs) {
+      VAST_ASSERT(rank(xs) > 0);
+      VAST_DEBUG(self, "got query for", rank(xs),
+                 "events in range [" << select(xs, 1) << ','
+                                     << (select(xs, -1) + 1) << ')');
+      auto requester = caf::actor_cast<receiver_type>(self->current_sender());
```
Comment (mavam) [inline libvast/src/system/archive.cpp:97]: I'm not sure if this is a safe operation. You cannot guarantee that the sender is `receiver_type`.

File: libvast/src/system/archive.cpp [hunk 3]
```diff
@@ -77,94 +77,124 @@ archive(archive_type::stateful_pointer<archive_state> self, path dir,
...
+      self->send(self, xs, requester);
+    },
+    [=](const ids& xs, receiver_type requester) {
+      auto& st = self->state;
+      if (st.active_exporters.count(requester->address()) == 0) {
+        VAST_DEBUG(self, "dismisses query for inactive sender");
```
Comment (mavam) [inline libvast/src/system/archive.cpp:103]: Should this ever happen? If not, I'd promote this to a warning?

File: libvast/src/system/archive.cpp [hunk 4]
```diff
@@ -77,94 +77,124 @@ archive(archive_type::stateful_pointer<archive_state> self, path dir,
...
+          }
+          t.stop(events);
+        },
+        [=](unit_t&, const error& err) {
+          if (err) {
+            VAST_ERROR(self, "got a stream error:", self->system().render(err));
+          }
+        });
+    },
+    [=](exporter_atom, const actor& exporter) {
```
Comment (mavam) [inline libvast/src/system/archive.cpp:177]: Why does the archive need to depend on the exporter?

File: libvast/src/system/archive.cpp (hunk 3 above)
Comment (dominiklohmann) [reply libvast/src/system/archive.cpp:103]: Yes, this does happen. We reviewed this in person.

File: libvast/src/system/archive.cpp (hunk 2 above)
Comment (dominiklohmann) [reply libvast/src/system/archive.cpp:97]: We talked about this in person: `receiver_type` defines a behavior that the caller must necessarily implement. I have added a check to see whether the cast succeeded.

**Comment by dominiklohmann** [comment]
Pivot is working now, please review again @mavam.

**Pull request merged**